### PR TITLE
Introduce land ice surface mass balance fields in default runs without MECs/GLC

### PR DESCRIPTION
**Pull request by chloewhicker**
Introduced new diagnostic Surface Mass Balance fields, they are calculated as if there is a coupled land ice component but are only diagnostic - it does not modify the energy/water budgets in default or coupled GLC simulations. 

[BFB]

-----

More info on modifications and testing can be found [here](https://acme-climate.atlassian.net/wiki/spaces/FAN/pages/4619599888/New+Feature+Overview+Document+ELM+Ice+SMB+h0+fields)

### CODE REVIEW

File: components/elm/src/biogeophys/HydrologyDrainageMod.F90 [hunk 1]
```diff
@@ -215,18 +217,30 @@ subroutine HydrologyDrainage(bounds,    &
...
+                   qflx_glcice_frz(c) = qflx_snwcp_ice(c)
+                   qflx_glcice(c) = qflx_glcice(c) + qflx_glcice_frz(c)
+                   if (glc_dyn_runoff_routing(g)) qflx_snwcp_ice(c) = 0._r8
          end if
+
+         !if (lun_pp%itype(l)==istice) then
+         !      qflx_glcice_frz_diag(c) = qflx_snwcp_ice(c)
+         !      qflx_glcice_diag(c) = qflx_glcice_diag(c) + qflx_glcice_frz_diag(c)
+         !endif
+
```
Comment (matthewhoffman) [inline components/elm/src/biogeophys/HydrologyDrainageMod.F90:243]: Is there any reason to keep this commented code?

File: components/elm/src/biogeophys/HydrologyDrainageMod.F90 [hunk 2]
```diff
@@ -215,18 +217,30 @@ subroutine HydrologyDrainage(bounds,    &
...
          ! In the following, we convert glc_snow_persistence_max_days to r8 to avoid overflow
          if ( (snow_persistence(c) >= (real(glc_snow_persistence_max_days, r8) * secspday)) &
-              .or. lun_pp%itype(l) == istice_mec) then
-            qflx_glcice_frz(c) = qflx_snwcp_ice(c)
-            qflx_glcice(c) = qflx_glcice(c) + qflx_glcice_frz(c)
-            if (glc_dyn_runoff_routing(g)) qflx_snwcp_ice(c) = 0._r8
+              .or. lun_pp%itype(l) == istice_mec ) then
+                   qflx_glcice_frz(c) = qflx_snwcp_ice(c)
+                   qflx_glcice(c) = qflx_glcice(c) + qflx_glcice_frz(c)
+                   if (glc_dyn_runoff_routing(g)) qflx_snwcp_ice(c) = 0._r8
```
Comment (matthewhoffman) [inline components/elm/src/biogeophys/HydrologyDrainageMod.F90:236]: Consider reverting these whitespace changes (though I see why you made them, so up to you)

File: components/elm/src/biogeophys/SoilTemperatureMod.F90 [hunk 3]
```diff
@@ -1643,8 +1646,7 @@ subroutine Phasechange_beta (bounds, num_nolakec, filter_nolakec, dhsdT, &
             ! as computed in HydrologyDrainageMod.F90.
 
             l = col_pp%landunit(c)
-            if (lun_pp%itype(l)==istice_mec) then
-
+            if ( lun_pp%itype(l)==istice_mec) then
```
Comment (matthewhoffman) [inline components/elm/src/biogeophys/SoilTemperatureMod.F90:1649]: Consider reverting this whitespace change

**Review by matthewhoffman — CHANGES_REQUESTED** [review]
@chloewhicker , thanks for this PR, and sorry for the delay in reviewing it.  I'm not familiar with this code, but I followed that you are mirroring the `qflx_glcice`  calculations with new `_diag` versions of the relevant variables, and that all makes sense.  I requested a number of code formatting/whitespace changes to make it clearer what the PR is actually doing.

File: components/elm/src/data_types/ColumnDataType.F90 [hunk 4]
```diff
@@ -5842,23 +5848,39 @@ subroutine col_wf_init(this, begc, endc)
      call hist_addfld1d (fname='QSNOFRZ', units='kg/m2/s', &
           avgflag='A', long_name='column-integrated snow freezing rate', &
            ptr_col=this%qflx_snofrz, set_lake=spval, c2l_scale_type='urbanf', default='inactive')
-
+
```
[the remark this answers is not in the document]
Comment (chloewhicker) [reply components/elm/src/data_types/ColumnDataType.F90:5851]: added this white space for consistency

File: components/elm/src/data_types/ColumnDataType.F90 [hunk 5]
```diff
@@ -5842,23 +5848,39 @@ subroutine col_wf_init(this, begc, endc)
...
-       this%qflx_glcice_frz(begc:endc) = spval
-       call hist_addfld1d (fname='QICE_FRZ',  units='mm/s',  &
-            avgflag='A', long_name='ice growth', &
-            ptr_col=this%qflx_glcice_frz, l2g_scale_type='ice')
-
-       this%qflx_glcice_melt(begc:endc) = spval
-       call hist_addfld1d (fname='QICE_MELT',  units='mm/s',  &
-            avgflag='A', long_name='ice melt', &
-            ptr_col=this%qflx_glcice_melt, l2g_scale_type='ice')
-    endif
```
[the remark this answers is not in the document]
Comment (chloewhicker) [reply components/elm/src/data_types/ColumnDataType.F90:5861]: this was just modified to include the else with the lines below - no whitespace change here

File: components/elm/src/biogeophys/BalanceCheckMod.F90 [hunk 6]
```diff
@@ -167,7 +167,7 @@ subroutine ColWaterBalanceCheck( bounds, num_do_smb_c, filter_do_smb_c, &
      use elm_varcon        , only : spval
      use column_varcon     , only : icol_roof, icol_sunwall, icol_shadewall
      use column_varcon     , only : icol_road_perv, icol_road_imperv
-     use landunit_varcon   , only : istice_mec, istdlak, istsoil,istcrop,istwet
+     use landunit_varcon   , only : istice_mec, istice, istdlak, istsoil,istcrop,istwet
```
[the remark this answers is not in the document]
Comment (chloewhicker) [reply components/elm/src/biogeophys/BalanceCheckMod.F90]: good call - removed

File: components/elm/src/biogeophys/HydrologyDrainageMod.F90 (hunk 1 above)
Comment (chloewhicker) [reply components/elm/src/biogeophys/HydrologyDrainageMod.F90]: no! good call

**Comment by chloewhicker** [comment]
Thanks for the feedback Matt - all comments have been addressed in the new commit!

**Review by matthewhoffman — APPROVED** [review]
@chloewhicker , thanks for making the minor adjustments from my review.  This looks good to me!  @bishtgautam , would you like to look this over?

File: components/elm/src/data_types/ColumnDataType.F90 [hunk 7]
```diff
@@ -5842,23 +5848,39 @@ subroutine col_wf_init(this, begc, endc)
...
+
+            this%qflx_glcice_frz(begc:endc) = spval
+             call hist_addfld1d (fname='QICE_FRZ',  units='mm/s',  &
+                  avgflag='A', long_name='ice growth', &
+                   ptr_col=this%qflx_glcice_frz, l2g_scale_type='ice')
+
+            this%qflx_glcice_melt(begc:endc) = spval
+             call hist_addfld1d (fname='QICE_MELT',  units='mm/s',  &
+                  avgflag='A', long_name='ice melt', &
+                   ptr_col=this%qflx_glcice_melt, l2g_scale_type='ice')
```
[the remark this answers is not in the document]
Comment (jonbob) [reply components/elm/src/data_types/ColumnDataType.F90]: @bishtgautam -- does that address your questions?

**Review by bishtgautam — APPROVED** [review]

**Comment by jonbob** [comment]
passes e3sm_developer on chrysalis except for a few I-cases that show DIFFs due to a change in fieldlists but otherwise BFB

merged to next

**Comment by wlin7** [comment]
The additional fields `QICE`, `QICE_FRZ`, `QICE_MELT` in elm.h0 do also cause prod_next tests to report DIFFs. The simulation results are otherwise unaffected.

**Comment by jonbob** [comment]
merged to master and expected DIFFs due only to different field lists blessed -- except for pm-cpu, which had issues overnight

**Pull request merged**